src/hotspot/share/prims/jvmtiTagMap.cpp

```diff
@@ -2224,6 +2226,159 @@ class JNILocalRootsClosure : public OopClosure {
   virtual void do_oop(narrowOop* obj_p) { ShouldNotReachHere(); }
 };
 
+// Helper class to collect/report stack references.
+class StackRefCollector {
+private:
+  JvmtiTagMap* _tag_map;
+  JNILocalRootsClosure* _blk;
+  // java_thread is needed only to report JNI local on top native frame;
+  // I.e. it's required only for platform/carrier threads or mounted virtual threads.
+  JavaThread* _java_thread;
+
+  oop _threadObj;
+  jlong _thread_tag;
+  jlong _tid;
+
+  bool _is_top_frame;
+  int _depth;
+  frame* _last_entry_frame;
+
+  bool report_java_stack_refs(StackValueCollection* values, jmethodID method, jlocation bci, jint slot_offset);
+  bool report_native_stack_refs(jmethodID method);
+
+public:
+  StackRefCollector(JvmtiTagMap* tag_map, JNILocalRootsClosure* blk, JavaThread* java_thread)
+    : _tag_map(tag_map), _blk(blk), _java_thread(java_thread),
+      _threadObj(nullptr), _thread_tag(0), _tid(0),
+      _is_top_frame(true), _depth(0), _last_entry_frame(nullptr)
+  {
+  }
+
+  bool set_thread(oop o);
+  // sets the thread and reports the reference to it with the specified kind.
```

Comment:
Could I ask you to polish comments a little bit?
Let's use the following Consistent Comment Rule (CCR):
If comment is started with a capitol letter then it should be ended with dot.
Otherwise, it should not be ended with dot.
I'll mark with CCR other comments that are inconsistent with CCR.